### PR TITLE
fix 39453

### LINKED ISSUE #39453: Call hierarchy: missing outgoing calls inside function expression
**TypeScript Version:**  3.9.2

**Search Terms:**

**Code**

```ts
const c = function () {
    return {};
};
const b = function () {
    return c();
};
const a = function () {
    return b();
};
a();
```

Place cursor on `a` (in `const a`) and then run "peek call hierarchy" (configured to show outgoing calls).

**Expected behavior:**

Call to `b` should be shown.

**Actual behavior:**

image

**Playground Link:** 

**Related Issues:** 

If we use function declarations instead of function expressions, this works as expected.

image

### PR DESCRIPTION
**Pull request by ShuiRuTian**
Fixes #39453

### CODE REVIEW

**Comment by ShuiRuTian** [comment]
@typescript-bot test this

**Comment by ShuiRuTian** [comment]
Emmmmm, is there something wrong? Why does it not listen to me?

**Comment by orta** [comment]
Only the compiler team can trigger it to avoid DOSing it 👍 

FWIW, `test this` isn't going to be useful here. It is only useful for code which changes the compiler's type checking etc because it runs your PR against a lot of real world codebases, which won't trigger the call hierarchy. [Docs are here](https://github.com/microsoft/TypeScript/wiki/Triggering-TypeScript-Bot)

File: src/services/callHierarchy.ts [hunk 1]
```diff
@@ -262,6 +262,12 @@ namespace ts.CallHierarchy {
                     }
                 }
             }
```
Comment (jessetrinity) [inline src/services/callHierarchy.ts:264]: Does the `else` make an actual difference here? Can this be reduced to the contained `if` check alone (or at least an `else if`)?

Comment (ShuiRuTian) [reply src/services/callHierarchy.ts:264]: It is OK to take the if clause out. But, not pretty sure, I think it split two conditions -- declaration and call expression.
```
function foo() {
    bar();
}

const bar = function () {    //declaration 
    baz();
}

function baz() {
}

bar()  // call expression
```
if followingSymbol is false and return, it means declaration, if followingSymbol is true, it means call expression.

I would have a close look at this when I have time.

Comment (ShuiRuTian) [reply src/services/callHierarchy.ts:264]: I test some cases, and think followingSymbol does mean whether is declaration or call expression. But some code are used for both condition to find returned value.
At least for the code I added, it is only used for the call expression of const function expression.
Do I still need to remove the else clause?

**Comment by DanielRosenwasser** [comment]
> Are we aware of a reason this should only work with `const` declared function expressions and not `let` or `var`?

We know that `const`s can never be assigned to - technically `function`s can be assigned to, but we expect that way less.

File: src/services/callHierarchy.ts (hunk 1 above)
Comment (jessetrinity) [reply src/services/callHierarchy.ts:264]: This comment might have shown up in the wrong place. You have an `else` on [line 265](https://github.com/microsoft/TypeScript/pull/39579/files#diff-46de2fd0324aef6e77de5b5b11a8239cR265) immediately followed by an `if`. Could you not do an `else if` here?

**Comment by ShuiRuTian** [comment]
@jessetrinity  Seems I did not explain clearly, sorry for that.
the code is like this:
```
while(true){
  if (condition1(location)) {
    return handleCode1(location);
  }
  if (condition2(location)) {
    return handleCode2(location);
  }

  if (!followingSymbol) {
    location = changeLocation();
    followingSymbol = true;
    continue;
  }
  return undefined;
}
```
So, what is the use of followingSymbol? I think it is to deal with two situations:
```
const /* trigger here */bar = function () {    // declaration 
}
/* trigger here */bar()  // call expression
```
If it is declaration situation, some handleCode would return in the first loop.
If it is call-expression situation, followingSymbol would be true, and ,as a result, some handleCode would return in the second loop.

Some handleCode code would work for the both situations. But, for the code I added, it would and should work only in call-expression situation, so I put them in the condition in which followingSymbol is true.

Is there still something I do not explain well this time?

Of course it is OK to remove the else clause. Do you wish me to do so after the explanation?

**Comment by jessetrinity** [comment]
@ShuiRuTian I think I'm finding it confusing because `followingSymbol` should be true in any case while in the second loop (otherwise `resolveCallHierarchyDeclaration` would just return `undefined`). If we return while in the first loop, I think that just means we were already at the variable declaration, so we did not need to "follow" the symbol. The `else` seems to imply that there is something special about `location` being a `VariableDeclaration`, but I don't see why that would be the case.

I would expect to see the new check above the `!followingSymbol` check with the others:

```ts
            ...
            if (isConstructorDeclaration(location)) {
                if (isValidCallHierarchyDeclaration(location.parent)) {
                    return location.parent;
                }
                return undefined;
            }
            if (isVariableDeclaration(location) && location.initializer && isConstNamedExpression(location.initializer)) {
                return location.initializer;
            }
            if (!followingSymbol) {
            ...
```

**Comment by ShuiRuTian** [comment]
Ah, got it!

**Comment by jessetrinity** [comment]
Awesome, thanks!

**Pull request merged**